gestalt-asset-core/src/main/java/org/terasology/gestalt/assets/Asset.java

```diff
@@ -50,12 +56,49 @@
 @API
 @ThreadSafe
 public abstract class Asset<T extends AssetData> {
+    private static final Logger logger = LoggerFactory.getLogger(Asset.class);
+
+    private final List<WeakReference<Asset<T>>> instances;
+    private Asset<T> parent;
 
     private final ResourceUrn urn;
     private final AssetType<?, T> assetType;
     private final DisposalHook disposalHook = new DisposalHook();
     private volatile boolean disposed;
 
+    private void appendInstance(Asset<T> asset) {
+        if (parent == null) {
+            Preconditions.checkArgument(!getUrn().isInstance());
+            asset.parent = this;
+            instances.add(new WeakReference<>(asset));
+        } else {
+            asset.parent = parent;
+            parent.instances.add(new WeakReference<>(asset));
+        }
+    }
+
+    protected Iterable<WeakReference<Asset<T>>> instances() {
+        if (this.parent != null) {
+            return parent.instances();
+        }
+        return instances;
+    }
+
+    protected void cleanup() {
```

Comment:
Can this be private? Since resource cleanup should happen in the disposal hook, what reason is there for implementors to change this?

Reply:
Or call it.